Add GL_MESA_EGL_sync extension

hfink-daqri opened the pull request:
This PR adds the `GL_MESA_EGL_sync` OpenGL extension to extend
`EGL_KHR_fence_sync` with OpenGL client API support.

While arguably OpenGL already has `GL_ARB_sync`, there are platform
extensions that are layered on top `EGL_KHR_fence_sync` rather than
`GL_ARB_sync`. Without OpenGL support of `EGL_KHR_fence_sync`, these
extension can't be used with an OpenGL command stream. One notable
example is the `EGL_ANDROID_native_fence_sync` extension, which enables
`linux_sync_file` objects to be used as the backend of EGL sync objects.
Linux sync file objects can be used to synchronize buffer access between
processes dispatching OpenGL drawing commands, or to synchronize with
the Linux KMS display subsystem (aka 'explicit fencing').

There is a corresponding Mesa merge-request [here](https://gitlab.freedesktop.org/mesa/mesa/merge_requests/624). An implemented use-case is demonstrated by the public [kms-quads](https://gitlab.freedesktop.org/daniels/kms-quads) sample code. Note that the Mesa MR originally attempted to directly expose `OES_EGL_sync` to GL core. As pointed out by feedback in the Mesa MR, this is not a viable path, since exposing another OpenGL ES extension in GL compat/core is breaking the CTS (`OES_EGL_image` being the only exception). In this pull request, we instead propose to add a simple Mesa extension that explicitly adds EGL fence sync support to OpenGL in a similar way as `OES_EGL_sync` does for OpenGL ES and `VG_KHR_EGL_sync` does for OpenVG.

Once the approach of this pull request is approved to be viable, I'll make sure the Mesa MR and the kms-quads sample code gets adapted accordingly.

oddhack commented:
@hfink-daqri same question as the other Mesa extension - is this a spec signed off by the Mesa project? I unfortunately don't recognize your name so I'd like to double-check this is OK with the project whose vendor tag it's using.

fooishbar commented:
@oddhack Still under discussion within Mesa; please hold off on merging.

pdaniell-nv commented:
@fooishbar Is this ready yet?

fooishbar commented:
@pdaniell-nv No, it's not ready. At this stage it's still an RFC, not something concrete and completed that should be merged. Is there any way we can better indicate the status?

pdaniell-nv commented:
> @pdaniell-nv No, it's not ready. At this stage it's still an RFC, not something concrete and completed that should be merged. Is there any way we can better indicate the status?

You can prefix the title with "WIP: ", which means work in progress.

fooishbar commented:
@pdaniell-nv Hi, this has now been reviewed and merged by Mesa, so we're ready to pull it in here.

pdaniell-nv commented:
> @pdaniell-nv Hi, this has now been reviewed and merged by Mesa, so we're ready to pull it in here.

Can you remove the "WIP" from the title.

hfink-daqri commented:
> Can you remove the "WIP" from the title.
@pdaniell-nv @fooishbar  done

oddhack commented:
@pdaniell-nv LGTM - just mark it Approved and I'll deal with it.

cubanismo commented:
Missed this somehow.  At the risk of appearing obstructionist by being so late, any interest in making this trivial extension an EXT before merging?  If there's urgency, no worries, but we'd probably be interested in exposing this as well, since I think we may actually be exposing the functionality already by mistake.

fooishbar commented:
@cubanismo Sure, more than happy to.

hfink-daqri commented:
@fooishbar @cubanismo I renamed the extension and updated the PR. Can you take a quick look if you are ok with it? Once merged, I'll go back to Mesa and change it there as well.

fooishbar commented:
@hfink-daqri LGTM as well. @pdaniell-nv / @oddhack can you please merge?

oddhack commented:
Merging per @fooishbar @cubanismo @pdaniell-nv .
